lib/teleterm/daemon/daemon.go

```diff
@@ -148,6 +153,35 @@ func (s *Service) createGateway(ctx context.Context, params CreateGatewayParams)
 		LocalPort:             params.LocalPort,
 		CLICommandProvider:    cliCommandProvider,
 		TCPPortAllocator:      s.cfg.TCPPortAllocator,
+		// TODO: Add test that makes sure NewWithLocalPort properly copies OnNewConnection.
+		OnNewConnection: func(gatewayURI uri.ResourceURI, targetURI string) {
+			// TODO Refactor targetURI from string to uri.ResourceURI.
+			// https://github.com/gravitational/teleport/issues/15953
+			clusterURI, err := uri.ParseClusterURI(targetURI)
+			if err != nil {
+				s.Log().WithError(err).Warnf(
+					"Malformed gateway target URI (%s), could not parse it into uri.ResourceURI",
+					targetURI,
+				)
+				return
+			}
+
+			clusterEvent := &api.ClusterEvent{
+				ClusterUri: clusterURI.String(),
+				Event: &api.ClusterEvent_NewGatewayConnectionAccepted{
+					NewGatewayConnectionAccepted: &api.NewGatewayConnectionAccepted{
+						GatewayUri: gatewayURI.String(),
+						TargetUri:  targetURI,
+					},
+				},
+			}
+
+			s.Log().Debug("Sending ClusterEvent to clusterEventsC")
+
+			s.clusterEventsC <- clusterEvent
```

Comment:
`clusterEventsC` is an unbuffered channel. If no stream is listening, handling of the connection can be blocked?

Reply:
Right, that's the problem I described in the "Making sure there's always a reader" section.

Using a buffer seemed to only postpone dealing with this problem, so I didn't opt for it.

In case of a recoverable stream error, we expect that the frontend app is going to establish a new stream ASAP. If that's not the case then we have a serious problem with the application.

With that in mind, perhaps it'd be a good idea to do this: inside the callback, spin up a new goroutine. The goroutine tries to write on the channel. If that is not possible within three seconds (that's an arbitrary amount of time), it disregards the message, logs the error and then quits.

---

OTOH I don't see why we'd want the connection to proceed when there's a fundamental problem with the app. As I said above, we expect any intermittent issues with tshd -> Connect connection to be resolved pretty quickly, so if it blocks on that write for a couple hundred of milliseconds then I think that's fine.

---

Now that I wrote all of this, I understand that in the case of an unrecoverable issue, we might actually want to stop this goroutine from being blocked. I suppose I could expose `net.Conn` to this function and close the connection if we can't write on the channel within some deadline. Does that sound good?

Reply:
Sorry, I missed that part in the description. I think it's ok to add `net.Conn` if that solves the situation. Personally, I would try to decouple the dependency of the stream when handling a new connection, and handle the problem of a broken stream elsewhere. Obviously, I am not familiar enough with all the flows so I don't know if there is a better place =p.